lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -674,83 +843,446 @@ export default function getAPIImplementation(): APIInterface {
+                                return signUpResponse;
+                            }
+
+                            signUpResponse.user = await linkAccountsForFactorSetup(
+                                sessionUser,
+                                signUpResponse.recipeUserId,
+                                userContext
+                            );
+
+                            await mfaInstance.recipeInterfaceImpl.markFactorAsCompleteInSession({
+                                session: session,
+                                factorId: "emailpassword",
+                                userContext,
+                            });
+
+                            return {
+                                status: "OK",
+                                session,
+                                user: signUpResponse.user,
+                            };
+                        }
+                    }
+                } catch (err) {
+                    if (err instanceof SignUpError) {
+                        return err.response;
+                    } else if (err instanceof RecurseError) {
+                        continue;
+                    } else {
+                        throw err;
+                    }
+                }
             }
+        },
+    };
+}
+
+class SignUpError extends Error {
+    response:
+        | {
+              status: "SIGN_UP_NOT_ALLOWED";
+              reason: string;
+          }
+        | {
+              status: "EMAIL_ALREADY_EXISTS_ERROR";
+          };
+
+    constructor(
+        response:
+            | { status: "SIGN_UP_NOT_ALLOWED"; reason: string }
+            | {
+                  status: "EMAIL_ALREADY_EXISTS_ERROR";
+              }
+    ) {
+        super(response.status);
 
-            // this function also does account linking
-            let response = await options.recipeImplementation.signUp({
+        this.response = response;
+    }
+}
+
+class SignInError extends Error {
+    response: {
+        status: "SIGN_IN_NOT_ALLOWED";
+        reason: string;
+    };
+
+    constructor(response: { status: "SIGN_IN_NOT_ALLOWED"; reason: string }) {
+        super(response.status);
+
+        this.response = response;
+    }
+}
+
+class RecurseError extends Error {
+    constructor() {
+        super("RECURSE");
+    }
+}
+
+const checkIfSignInIsAllowed = async (tenantId: string, user: User, userContext: UserContext) => {
+    let isSignInAllowed = await AccountLinking.getInstance().isSignInAllowed({
+        user,
+        tenantId,
+        userContext,
+    });
+
+    if (!isSignInAllowed) {
+        throw new SignInError({
+            status: "SIGN_IN_NOT_ALLOWED",
+            reason:
+                "Cannot sign in due to security reasons. Please try resetting your password, use a different login method or contact support. (ERR_CODE_008)",
+        });
+    }
+};
+
+const checkIfValidFirstFactor = async (tenantId: string, userContext: UserContext) => {
```

Comment:
Move this inline the API cause it assumes no session exists.

Reply:
Done